src/main/src/main/java/org/geoserver/security/impl/DefaultResourceAccessManager.java

```diff
@@ -192,85 +207,91 @@ public boolean canAccess(Authentication user, ResourceInfo resource, AccessMode
         }
         
         // if we have a catalog rule that is at resource level, it's the most specific type,
-        // it wins. As an alternative, it could be that we do not need to check layer groups at all
-        SecureTreeNode catalogNode = root.getDeepestNode(new String[] { workspace, resourceName });
-        int catalogNodeDepth = catalogNode.getDepth();
-        final boolean catalogNodeAllowsAccess = catalogNode.canAccess(user, mode);
+        // it wins. Or it could be that we do not need to check layer groups at all
+        SecureTreeNode securityNode = root.getDeepestNode(new String[] { workspace, resourceName });
+        int catalogNodeDepth = securityNode.getDepth();
+        boolean rulesAllowAccess = securityNode.canAccess(user, mode);
         if(catalogNodeDepth == SecureTreeNode.RESOURCE_DEPTH || !layerGroupContainmentCheckRequired()) {
-            return catalogNodeAllowsAccess;
+            return rulesAllowAccess;
         }
-
-        // check if there is any containing group that has security rules, if not, the deepest
-        // catalog rule wins
-        List<SecureTreeNode> nodes = getNodesContainingResource(resource);
-        if(nodes.isEmpty()) {
-            return catalogNodeAllowsAccess;
+        
+        // grab the groups containing the resource, if any. If none, there is no group related logic to apply
+        Collection<LayerGroupSummary> containers  = groupsCache.getContainerGroupsFor(resource);
+        if(containers.isEmpty()) {
+            return rulesAllowAccess;
         }
         
-        // if we have at least one containing tree that is secured and authorizes access, then 
-        // we have a winner
-        if(nodes.stream().anyMatch(n -> n.canAccess(user, mode) && n.getDepth() > catalogNodeDepth)) {
-            return true;
+        // there are groups, so there might be more specific rules overriding the catalog one, search for them
+        List<LayerGroupSummary> groupOverrides = containers.stream().filter(sg -> {
+            LayerGroupInfo gi = rawCatalog.getLayerGroup(sg.getId());
+            if(gi == null) {
+                return false;
+            }
+            SecureTreeNode node = getNodeForGroup(gi);
+            return (node != null && node.getDepth() > catalogNodeDepth) || (sg.getMode() == Mode.OPAQUE_CONTAINER);
```

Comment:
This lines confuses me a bit, I don't understand why we have the ``sg.getMode() == Mode.OPAQUE_CONTAINER`` and the node depth check at the same time.